### ISSUE TITLE
Unable to Export, "connect ECONNREFUSED 127.0.0.1:24981"

### ISSUE THREAD

**Issue by BattouSaeen** [Raider-Arts/painter-megascan-link#2]
Hello,

Thanks for this plugin! However, I am unable to export.

Log lists error in title.
I followed a forum post to install some dependencies via cmd:
```
C:\Program Files\Allegorithmic\Substance Painter\resources\pythonsdk>python.exe -m pip websocket-client
Traceback (most recent call last):
  File "C:\Program Files\Allegorithmic\Substance Painter\resources\pythonsdk\lib\runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "C:\Program Files\Allegorithmic\Substance Painter\resources\pythonsdk\lib\runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "C:\Program Files\Allegorithmic\Substance Painter\resources\pythonsdk\lib\site-packages\pip\__main__.py", line 16, in <module>
    from pip._internal import main as _main  # isort:skip # noqa
  File "C:\Program Files\Allegorithmic\Substance Painter\resources\pythonsdk\lib\site-packages\pip\_internal\__init__.py", line 19, in <module>
    from pip._vendor.urllib3.exceptions import DependencyWarning
  File "C:\Program Files\Allegorithmic\Substance Painter\resources\pythonsdk\lib\site-packages\pip\_vendor\urllib3\__init__.py", line 7, in <module>
    from .connectionpool import (
  File "C:\Program Files\Allegorithmic\Substance Painter\resources\pythonsdk\lib\site-packages\pip\_vendor\urllib3\connectionpool.py", line 7, in <module>
    from socket import error as SocketError, timeout as SocketTimeout
  File "C:\Program Files\Allegorithmic\Substance Painter\resources\pythonsdk\lib\socket.py", line 49, in <module>
    import _socket
ImportError: Module use of python38.dll conflicts with this version of Python.
```
Now is this because of any connection settings? I have a couple different versions of Python installed as well, which I refer to with the last line of there being some conflict. My current in-use version is 3.9.0. Other things over localhost work fine, so I'm guessing something to do with python version, but I am not experienced enough to fiddle with that without guidance.

**Comment by faggionluca** [comment]
Hello!
I suspect there is a conflict with your installed version/s of python (do you use any virtualenv (pipenv, conda, etc..)? or I better should ask you, how do you manage them?).

The first thing you should try is, if they are referenced in a **PATH (or PYTHONPATH) environment variable** just remove the entry from the list and try again because I suspect that pip is trying to install the websocket-client in your current python installation and not the one that ship with designer.

Let me know if this was helpful 😄

**Comment by BattouSaeen** [comment]
Hey! Nope, don't use any of those, don't even know what they do haha. I do however use a few things that might be using them on their own. For instance, the app Gramps uses python, SVP 4 uses python (there is an environment variable in pythonpath that leads to the executable of SVP 4), but I removed all other path variables leading to the python directory (curiously there was one pointing to C:\Python27). But none of it made any difference.

While I'm on the subject, what's the difference between the top "User variables for <username>" and the bottom "System variables" lists? The obvious meaning seems to be system-wide and local-only variables, but is there anything else to it? Especially since both sections had "paths" leading to python (and likewise for other programs).

**Comment by faggionluca** [comment]
Virtualenvs are just a way to self contain multiple python installation each with it's own set of dependencies (python module) that you can switch between them (in other words you can switch between different sets of modules and python versions combinations) but anyway regarding the PATH issue if you can restore the paths you deleted just to not harm your program installations

Try to run this command in a command prompt and report the output:
``` which python ```

Regarding the last question, the difference between them is the visibility regarding the current logged user in the system the User variables are specific for each user so multiple users can have different variables meanwhile the all see and share the Global variables

**Comment by BattouSaeen** [comment]
"Which python" spat an error out. If you're asking for version, I can use "py --version". Curiously, if I use "python --version" or "python.exe --version", Windows will show its modern message of "this app cannot run on this PC". Is this because I removed the paths?

--version brings up 3.9.0 btw.

**Comment by faggionluca** [comment]
**I'm so sorry** I was supposed to write **``where python``** instead of which
This command will list the path where windows found a python.exe executable so we can see if in the PATH variables there are still listed additional python installation and anyway just a small advice when you delete a path from an enviroment variable make ALWAYS a backup (store the string of the path in a file or anything)

**Comment by BattouSaeen** [comment]
Ah I see.
So I have the following results from which python:

C:\WINDOWS\System32\python.exe
C:\Users\username\AppData\Local\Programs\Python\Python39\python.exe
C:\Program Files (x86)\SVP 4\mpv64\python.exe

🤪😵

**Comment by faggionluca** [comment]
Okay since you have a lot of installation (even system ones) remove them is not feasible
Lets try to inform the correct version of python to pip by adding the ```C:\Program Files\Allegorithmic\Substance Painter\resources\pythonsdk``` path to the PATH environment variable before anything else (at the top of the list)
and then retry to execute the command ```python.exe -m pip websocket-client```  from the ```C:\Program Files\Allegorithmic\Substance Painter\resources\pythonsdk``` path and lets hope for the best 🤞

**Comment by BattouSaeen** [comment]
Unfortunately didn't work either :(

**Comment by faggionluca** [comment]
This is my last attempt since I'm running out of ideas, instead of putting the path in the PATH variable put it in the PYTHONPATH variable if you don't have it in the list simple create a new variable and name it PYTHONPATH, (ah by the way every time you update an environment variable you should close and reopen the terminal if it was open otherwise the terminal wont see the changes you just made, eheh silly windows 😄 )

**Comment by BattouSaeen** [comment]
Okay that worked! The CMD part completed, but export is still failing.

In Substance Painter's log, I saw the same error about conflicts. If I start or stop the plugin (python plugin) then it errors that it can neither start nor stop the plugin.

![image](https://user-images.githubusercontent.com/4526700/101205866-1ef05b80-36d3-11eb-9d9e-d0780f15a7d8.png)

**Comment by faggionluca** [comment]
Okay nice to hear that part worked, now moving on Painter try to add the path ```C:\Program Files\Allegorithmic\Substance Painter\resources\pythonsdk ``` also to the PATH environment variable and restart painter

**Comment by BattouSaeen** [comment]
Okay, it worked! Thanks a lot! Think you can summarise what was happening? In particular, how should be managing all these python installations since many apps use python. Shouldn't they all use one python and just have config folders in their own local directories for their business but use the main instance of python, instead of installing a copy for themselves? What should the path and pythonpath have as the default value for when python needs to be used "globally", instead of a specific path to a specific app's folder? I do have a programming background but it's not very experienced, especially when it comes to python. So feel free to ELI5 it but without being too strict with dumbing it down.

**Comment by faggionluca** [comment]
Honestly I don't know all the exact nitty gritty details regarding python environment scanning dependencies gathering etc.. (but after this I'll surely going to look further into this subject) but what can I can give you some advices I gathered in my experience programming with python, first thing I can suggest you is looking into a virtualenv manager so you don't clutter the environment path I personally use [Anaconda ](https://www.anaconda.com/products/individual).

So instead of installing multiple pythons I simple create multiple virtualenv in anacoda and when I want to use one on them i simply write in the prompt ``` conda activate virtualenvname ``` and after that Anaconda will populate all the variables with the correct values and the python command is pointing to the correct python executable, for the part regarding the management of app bundled python version it's not something you "user" shouldn't worry about but even for this reason I use conda to avoid conflicts of this genre.

I'm sorry that I couldn't elaborate more on the subject and if you felt that I somewhat treated you like someone that don't know much about programming I didn't that on purpose of making you fell stupid or something but I do this to better assist people so I don't miss giving others information that I take for granted

I hope this is enough for you 😄

**Comment by BattouSaeen** [comment]
Awesome, thanks a lot! Sounds like something like conda is what I was asking for. And no way, this is the best way to explain it, I prefer people explaining things thoroughly; I do the same when explaining something. I only mentioned that I have programming background so you wouldn't have to dumb it down too much (since it takes time to explain things after all!). So thanks so much again! :D
